ts/features/pn/store/sagas/watchPaymentStatusSaga.ts

```diff
@@ -0,0 +1,247 @@
+      // un-observed payment to the map
+      if (paymentsToTrackMap.size <= maxVisiblePaymentsInDetailsUI) {
+        const paymentId = updatePaymentForMessageAction.payload.paymentId;
+        paymentsToTrackMap.set(paymentId, O.none);
+      }
+    } else if (
+      isActionOf(
+        updatePaymentForMessage.success,
+        updatePaymentForMessageAction
+      ) ||
+      isActionOf(updatePaymentForMessage.failure, updatePaymentForMessageAction)
+    ) {
+      const paymentId = updatePaymentForMessageAction.payload.paymentId;
+      // This check is not really be necessary but it is here to strengthen the solution
+      if (paymentsToTrackMap.has(paymentId)) {
+        yield* call(
+          addPaymentStatusToMap,
+          paymentId,
+          updatePaymentForMessageAction,
+          paymentsToTrackMap
+        );
+        const trackedPaymentsArray = Array.from(paymentsToTrackMap.values());
+        // We are done processing as soon as all payments have been updated.
+        // We know we are not ready yet as soon as we find a payment that
+        // does not have an update (O.none)
+        const isDoneProcessing = !trackedPaymentsArray.some(
+          maybeProcessedPayment => O.isNone(maybeProcessedPayment)
+        );
+        if (isDoneProcessing) {
+          // All payments' update have been retrieved, compute, track and end
+          yield* call(
+            computeAndTrackPaymentStatuses,
+            messageId,
+            trackedPaymentsArray
+          );
+          return;
+        }
+      }
+    }
+  }
+}
+
+function addPaymentStatusToMap(
+  paymentId: string,
+  action:
+    | ActionType<typeof updatePaymentForMessage.success>
+    | ActionType<typeof updatePaymentForMessage.failure>,
+  paymentsToTrack: Map<string, O.Option<PaymentStatus>>
+) {
+  if (isActionOf(updatePaymentForMessage.success, action)) {
+    paymentsToTrack.set(paymentId, O.some(payablePayment));
+  } else {
+    const details = action.payload.details;
+    paymentsToTrack.set(paymentId, O.some(processedPayment(details)));
+  }
+}
+
+function* computeAndTrackPaymentStatuses(
+  messageId: UIMessageId,
+  paymentStatuses: ReadonlyArray<O.Option<PaymentStatus>>
+) {
+  const currentFiscalCode = yield* select(profileFiscalCodeSelector);
+  const message = yield* select(state =>
+    pnMessageFromIdSelector(state, messageId)
+  );
+  const payments = yield* call(
+    paymentsFromPNMessagePot,
+    currentFiscalCode,
+    message
+  );
+
+  const paymentCount = payments?.length ?? 0;
+  const partialPaymentStatistics = paymentStatuses.reduce(
+    (accumulator, paymentStatus) =>
+      pipe(
+        paymentStatus,
+        computePartialPaymentStatistics,
+        sumPaymentStatistics(accumulator)
+      ),
+    partialPaymentStatusStatistics({})
+  );
+  const paymentStatusStatisticsObj = yield* call(
+    paymentStatusStatistics,
+    paymentCount,
+    partialPaymentStatistics
+  );
```

Comment:
What if we changed the code like this?

```suggestion
  const partialPaymentStatistics = paymentStatuses.reduce(
    (accumulator, maybePaymentStatus) =>
      pipe(
        maybePaymentStatus,
        O.map(paymentStatus =>
          pipe(
            paymentStatus,
            foldPaymentStatus(
              () => ({
                ...accumulator,
                unpaidCount: accumulator.unpaidCount + 1
              }),
              details => {
                switch (details) {
                  case Detail_v2Enum.PAA_PAGAMENTO_DUPLICATO:
                  case Detail_v2Enum.PPT_PAGAMENTO_DUPLICATO:
                    return {
                      ...accumulator,
                      paidCount: accumulator.paidCount + 1
                    };

                  case Detail_v2Enum.PAA_PAGAMENTO_SCADUTO:
                    return {
                      ...accumulator,
                      expiredCount: accumulator.expiredCount + 1
                    };

                  case Detail_v2Enum.PAA_PAGAMENTO_ANNULLATO:
                    return {
                      ...accumulator,
                      revokedCount: accumulator.revokedCount + 1
                    };

                  case Detail_v2Enum.PAA_PAGAMENTO_IN_CORSO:
                  case Detail_v2Enum.PPT_PAGAMENTO_IN_CORSO:
                    return {
                      ...accumulator,
                      ongoingCount: accumulator.ongoingCount + 1
                    };

                  default:
                    return accumulator;
                }
              }
            )
          )
        ),
        O.getOrElse(() => accumulator)
      ),
    {
      unpaidCount: 0,
      paidCount: 0,
      errorCount: 0,
      expiredCount: 0,
      revokedCount: 0,
      ongoingCount: 0
    }
  );

  yield* call(trackPNPaymentStatus, {
    paymentCount,
    ...partialPaymentStatistics
  });
```

Reply:
Yes, it is more concise and maybe readable. The only thing that I would like to keep is the using of utils functions instead of the switch (`ts/utils/payment.ts`) in order not to duplicate the logic that distinguish between processed payment categories. What do you think about this one?

Reply:
Okay, sounds good

Reply:
Applied [d961e10](https://github.com/pagopa/io-app/pull/5198/commits/d961e10b867bca878f732a3d85ba563ddc884c1a)